### PR TITLE
pending transactions, no test yet, too terrified to run the code

### PR DESCRIPTION
**Pull request by antonbobkov**
:scream: :scream: :scream: :scream: :scream:

### CODE REVIEW

**Comment by antonbobkov** [comment]
we should chat before you look at that code

File: src/pending_transactions.erl [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+
+%% update counters in order at all nonces in TxMap starting at NonceStart
+
+update_all_the_counters(NonceStart, Counter, TxMap) -> 
+
+    NonceList = lists:sort(lists:filter(fun (N) -> N >= NonceStart end, maps:keys(TxMap))),
+    
+    {CounterNew, TxMapNew} = lists:foldl(fun update_counter_at_nonce/2, {Counter, TxMap}, NonceList),
+    {CounterNew, TxMapNew}.
```
Comment (antonbobkov) [inline src/pending_transactions.erl]: ad hoc mapmapaccum

File: src/pending_transactions.erl [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    end,
+
+    if Status == ignore_duplicate ->
+	    PendingTx;
+       Status /= ignore_duplicate ->
+	    {CounterNew, TxMapNew} = insert_transaction(Nonce, Transaction, Counter, TxMap),
+
+	    update_container(PlayerId, TxMapNew, CounterNew, PendingTx)
+    end.
+
```
Comment (antonbobkov) [inline src/pending_transactions.erl]: forgot to return status alongside new PendingTx

oh well

File: src/pending_transactions.erl [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+%% @doc Initialize the container.
+
+init() ->
+    #pending_tx{player_map = maps:new(), counter = 0}.
+
+%% update Pending Transaction container with a new counter and a new transaction map for a player
+
+update_container(PlayerId, TxMp, NewCounter, PendingTx) -> 
+    %% erlang too hard
```
Comment (pdlla) [inline src/pending_transactions.erl]: 😱

File: src/pending_transactions.erl [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+-include("potato_records.hrl").
+
+%% @doc Initialize the container.
+
+init() ->
+    #pending_tx{player_map = maps:new(), counter = 0}.
+
+%% update Pending Transaction container with a new counter and a new transaction map for a player
+
```
Comment (pdlla) [inline src/pending_transactions.erl]: try -spec :D

File: src/pending_transactions.erl [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+update_container(PlayerId, TxMp, NewCounter, PendingTx) -> 
+    %% erlang too hard
+
+    PM = (PendingTx#pending_tx.player_map)#{PlayerId => TxMp},
```
Comment (pdlla) [inline src/pending_transactions.erl]: what does PlayerId => TxMp do?

Comment (pdlla) [reply src/pending_transactions.erl]: or what does this line of code do 😱

Comment (pdlla) [reply src/pending_transactions.erl]: oh nevermind I figured it out.

Maybe rename to add_playertxs_to_container or something like that?

Comment (pdlla) [reply src/pending_transactions.erl]: I guess I could have read the comments 😱

File: src/pending_transactions.erl [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    
+    PendingTx#pending_tx{player_map = PM, counter = NewCounter}.
+
+
+%% udpate a counter at a TxMap[Nonce] and return modified map and incremented counter
+
+update_counter_at_nonce(Nonce, {Counter, TxMap}) ->
+    {_OldCounter, Data} = maps:get(Nonce, TxMap),
```
Comment (pdlla) [inline src/pending_transactions.erl]: such _

File: src/pending_transactions.erl (hunk 2 above)
Comment (pdlla) [reply src/pending_transactions.erl]: w/e

File: src/pending_transactions.erl [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+    TxMapNew = maps:put( {Counter, Data}, TxMap),
```
Comment (antonbobkov) [inline src/pending_transactions.erl]: should be maps:put(Nonce, {Counter, Data}, TxMap)

File: src/pending_transactions.erl [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    {Counter + 1, TxMapNew}.
+
```
Comment (antonbobkov) [inline src/pending_transactions.erl]: should be maps:put(Nonce, {Counter, Data}, TxMap)

Comment (pdlla) [reply src/pending_transactions.erl]: 😱

**Comment by antonbobkov** [comment]
counter starts at zero and automatically increments on its own

what do you mean "How do you pick the first transaction to include out of the txmap?"

**Comment by pdlla** [comment]
potato mapaccum using fold LGTM 👍

**Comment by pdlla** [comment]
I guess I just don't even know what the counter is for.

**Comment by antonbobkov** [comment]
each tx will have a unique counter, some counters are skipped

counters determine the order of transactions to put into a new block

File: src/pending_transactions.erl [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+	    ok;
+	error ->
+	    Status = added_new
+    end,
+
+    if Status == ignore_duplicate ->
+	    PendingTx;
+       Status /= ignore_duplicate ->
+	    {CounterNew, TxMapNew} = insert_transaction(Nonce, Transaction, Counter, TxMap),
+
```
Comment (pdlla) [inline src/pending_transactions.erl]: what about status == added_new?

**Comment by pdlla** [comment]
Like if you are making a block and you want to add a tx from pending tx, how do you get them?

And how do you pick what counter value to call insert_transaction with

**Comment by antonbobkov** [comment]
you'd never call insert_transaction directly, you'd call add_transaction :scream: 

new block transactions are given by
```
get_pending_transactions(_, _) ->
    ok.
```

:scream: :scream:

**Comment by antonbobkov** [comment]
(TBD)

**Comment by pdlla** [comment]
> you'd never call insert_transaction directly, you'd call add_transaction 😱
> 
> new block transactions are given by
> 
> ```
> get_pending_transactions(_, _) ->
>     ok.
> ```
> 
> 😱 😱

🤔

**Comment by pdlla** [comment]
Who decides what counter inside the PendingTx record that you call add_transaction with?

**Comment by antonbobkov** [comment]
it starts at zero and automatically increments

**Comment by pdlla** [comment]
Will the counter thing make it easy for a single player to spam a whole bunch of transactions and fill up the block? What about round robin through all the players?

**Comment by antonbobkov** [comment]
its like a private variable to keep track of order internally

**Comment by pdlla** [comment]
> its like a private variable to keep track of order internally

oh I get it now 😂

**Comment by antonbobkov** [comment]
they could do that (spam) anyway without a counter

**Comment by pdlla** [comment]
Ok, this counter accumulator thingy is super cool!!

But I think round robin through all the accounts might be better 😱 

Since the players in a game our permissioned, you can't do sybil attacks and every player gets allocated the same number of tx/block. I guess it could be a little weird if you have a reorg where one player has a ton of transactions that go back to pending.

**Pull request merged**